incremental.go

```diff
@@ -0,0 +1,267 @@
+	bucket, err := bfs.Connect(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := NewIncrementalProducerForBucket(ctx, bucket, opt, modfn, pfn)
+	if err != nil {
+		_ = bucket.Close()
+		return nil, err
+	}
+	p.ownBucket = true
+
+	return p, nil
+}
+
+// NewIncrmentalProducerForRemote starts a new incremental feed producer for a bucket.
+func NewIncrementalProducerForBucket(ctx context.Context, bucket bfs.Bucket, opt *IncrementalProducerOptions, modfn LastModifiedFunc, pfn IncrementalProduceFunc) (*IncrementalProducer, error) {
+	var o IncrementalProducerOptions
+	if opt != nil {
+		o = *opt
+	}
+	o.norm()
+
+	ctx, stop := context.WithCancel(ctx)
+	p := &IncrementalProducer{
+		bucket: bucket,
+		ctx:    ctx,
+		stop:   stop,
+		opt:    o,
+		modfn:  modfn,
+		pfn:    pfn,
+	}
+
+	// run initial push
+	if _, err := p.push(); err != nil {
+		_ = p.Close()
+		return nil, err
+	}
+
+	// start continuous loop
+	// use error group
+	go p.loop()
+
+	return p, nil
+}
+
+// Close stops the producer.
+func (p *IncrementalProducer) Close() error {
+	p.stop()
+	if p.ownBucket {
+		return p.bucket.Close()
+	}
+	return nil
+}
+
+// LastPush returns time of last push attempt.
+func (p *IncrementalProducer) LastPush() time.Time {
+	return timestamp(atomic.LoadInt64(&p.lastPush)).Time()
+}
+
+// LastModified returns time at which the remote feed was last modified.
+func (p *IncrementalProducer) LastModified() time.Time {
+	return timestamp(atomic.LoadInt64(&p.lastMod)).Time()
+}
+
+// NumWritten returns the number of values produced during the last push.
+func (p *IncrementalProducer) NumWritten() int {
+	return int(atomic.LoadInt64(&p.numWritten))
+}
+
+func (p *IncrementalProducer) loadManifest() (*Manifest, error) {
+	remote := bfs.NewObjectFromBucket(p.bucket, "manifest.json")
+	defer remote.Close()
+
+	m := new(Manifest)
+
+	r, err := NewReader(p.ctx, remote, nil)
+	if errors.Is(err, bfs.ErrNotFound) {
+		return m, nil
+	} else if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	if err := r.Decode(m); errors.Is(err, bfs.ErrNotFound) { // some BFS implementations defer Open-ing the S3 object till first Decode call
+		return m, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+func (p *IncrementalProducer) loop() {
+	ticker := time.NewTicker(p.opt.Interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-ticker.C:
+			state, err := p.push()
+			if p.opt.AfterPush != nil {
+				p.opt.AfterPush(state, err)
+			}
+		}
+	}
+}
+
+func (p *IncrementalProducer) push() (*IncrementalProducerPush, error) {
+	start := time.Now()
+	atomic.StoreInt64(&p.lastPush, timestampFromTime(start).Millis())
+
+	// get last mod time for local records
+	localLastMod, err := p.modfn(p.ctx)
+	if err != nil {
+		return nil, err
+	}
+	if localLastMod.IsZero() {
+		return &IncrementalProducerPush{IncrementalProducer: p}, nil
```

Comment:
why are we returning if there are no existing records (e.g. on an initial push)?

Reply:
ok.  I assumed if we have no records, we would not need to push anything...  would we just write an empty data file? and last mod in manifest is time.Now()?

Reply:
I guess this is needed if we delete all records too...

Reply:
oh, sorry, that's local, yes, ignore!